### PR TITLE
Add RayJob CRD and controller logic

### PR DESCRIPTION
**Pull request by harryge00**
## Why are these changes needed?
Add RayJob CRD and its controller. 
Includes:
1. Get or create corresponding Rayclusters for ray jobs.
2. Connect ray dashboard
3. Submit ray job and check job status.

This PR borrows a lot from @brucez-anyscale , thanks so much.

## Related issue number

#106 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Jeffwan** [comment]
Thanks for the contribution! I will have a check today

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:27]: is the entrypoint the only required field? Rest of them are optional?

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:29]: `rayClusterConfig` do we want to use different name here?

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:28]: `job_id` -> Let's be consistent and use camel case `jobId` here?

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:38]: Do we need jobId and clusterName here?

File: ray-operator/config/crd/bases/ray.io_rayservices.yaml
Comment (Jeffwan) [inline ray-operator/config/crd/bases/ray.io_rayservices.yaml:54]: em. Did we forget to generate manifest after introducing `imagePullPolicy`?

File: ray-operator/config/crd/patches/cainjection_in_ray_rayjobs.yaml
Comment (Jeffwan) [inline ray-operator/config/crd/patches/cainjection_in_ray_rayjobs.yaml:2]: I fee we don't necessarily need these patches so we don't need to check them in? WDYT?

File: ray-operator/config/samples/ray_v1alpha1_rayjob.yaml
Comment (Jeffwan) [inline ray-operator/config/samples/ray_v1alpha1_rayjob.yaml:1]: Can we add an example here?

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:15]: is it `/api/jobs`?

File: ray-operator/controllers/ray/rayservice_controller.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/rayservice_controller.go:211]: nice refactor!

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/rayjob_controller.go:82]: minor: I feel the time we initialize the client, it should be ready to use. Seems there's no need to have two lines here?

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/rayjob_controller.go:85]: Typo?  -> `checkIfNeedSubmitJob`

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/rayjob_controller.go:103]: Let's change to use `jobStatus`

**Review by Jeffwan — COMMENTED** [review]
Seems some core logic in job controller is not ready yet. I left some comments and I will come back later for another round of review.

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (brucez-anyscale) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:66]: Is it better to keep this file only for CRD related and move other struct into controller package?

File: ray-operator/config/crd/patches/webhook_in_ray_rayjobs.yaml
Comment (brucez-anyscale) [inline ray-operator/config/crd/patches/webhook_in_ray_rayjobs.yaml:2]: may not need this for now

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:56]: is this log from RayCluster?

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:69]: just return err, since we will anyway create it

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/rayjob_controller.go:85]: +1 for rename

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:88]: rename

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:95]: check job status?

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:109]: is this the job retry logic?

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:108]: This part is not needed. You will reuse the same cluster. What you need is to terminate the job and start the job again in the same cluster. I do not think RayJob should terminate a cluster

**Review by brucez-anyscale — COMMENTED** [review]
Thanks. Please update the naming in job controller and do some manual integration test to ensure it works.

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:121]: do we need to watch RayCluster?

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:146]: here we also need to consider if the status update fails, do we create multiple clusters.

File: ray-operator/controllers/ray/utils/dashboard_httpclient_test.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/utils/dashboard_httpclient_test.go:18]: Great to have http client tests.

**Comment by Jeffwan** [comment]
@harryge00 Did you get a chance to update the PR. We are close to v0.3.0 release and we plan to include this feature in the release.

**Comment by harryge00** [comment]
@Jeffwan @brucez-anyscale  I have just updated this PR, could u take another review?

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (harryge00) [reply ray-operator/controllers/ray/rayjob_controller.go]: I agree with you. I saw you use two status (pending/active cluster) to record. I may submit another PR to resolve this issue

**Comment by Jeffwan** [comment]
@harryge00 Could you help fix the linter and build issue?

**Comment by harryge00** [comment]
> @harryge00 Could you help fix the linter and build issue?
I have fixed the linter and build issue.
I can pass the compatibility test locally. How can trigger the compatibility test to run again? 
@Jeffwan

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/rayjob_controller.go:66]:
```suggestion
	// Get RayJob instnace
```

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:13]: Let's remove `_` for consistency? `JobStatusPending` should be good

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayjob_types.go:23]: same here

File: ray-operator/controllers/ray/common/service.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/common/service.go:23]: Let's use `klog` for consistency

File: ray-operator/controllers/ray/common/service.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/common/service.go:24]: TODO item will be done in a separate logic when raycluster_controller reconcile services.

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (Jeffwan) [reply ray-operator/controllers/ray/rayjob_controller.go:64]: Yes. https://github.com/ray-project/kuberay/blob/68faa10bc4af44c565fda9fab731f0de0935287e/ray-operator/controllers/ray/raycluster_controller.go#L40

It should be changed to Logger of current controller. BTW, I will make some changes to remove package scope `log` in raycluster. It's kind of annoying.

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (Jeffwan) [reply ray-operator/controllers/ray/rayjob_controller.go:94]: we probably should use metadata to customize job id (using kubernetes job name) then we can retrieve the job from the cluster to see if there's existing one

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:267]: klog

**Comment by Jeffwan** [comment]
I think the polished version is close to complete. Great work! Can you help address my above comments @harryge00

File: ray-operator/apis/ray/v1alpha1/rayjob_types.go
Comment (brucez-anyscale) [reply ray-operator/apis/ray/v1alpha1/rayjob_types.go:13]: +1

File: ray-operator/config/crd/patches/cainjection_in_ray_rayjobs.yaml
Comment (brucez-anyscale) [inline ray-operator/config/crd/patches/cainjection_in_ray_rayjobs.yaml:4]: Pls delete some unnecessary files.

File: ray-operator/config/rbac/role.yaml
Comment (brucez-anyscale) [inline ray-operator/config/rbac/role.yaml:8]: is this auto generated?

File: ray-operator/controllers/ray/common/service.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/common/service.go:17]: Please check the `ray-operator/controllers/ray/rayservice_controller.go:442`. Do you need dashboard url or dashboard agent url? I think job submit still the `dashboard url`

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/rayjob_controller.go:64]: Pls fix this and see how RayService uses its own logger

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/rayjob_controller.go:94]: Still the same comment on this check

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/rayjob_controller.go:165]: better split saving cluster name into CR state and create Ray Cluster into 2 reconcile loop. Otherwise, if CR state update fails, you will create another RayCluster, cause resource leak.

Comment (brucez-anyscale) [reply ray-operator/controllers/ray/rayjob_controller.go:165]: You can check how RayService handles it

**Comment by brucez-anyscale** [comment]
We should at least add unit tests. Also please manually test it in EKS or GKE or kind cluster.

**Comment by Jeffwan** [comment]
@harryge00 Please help fix the build issue and lint issues. The logics look good to me and we can file follow up PRs to address potential issue. 

image

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (harryge00) [inline ray-operator/controllers/ray/rayjob_controller.go:76]: @brucez-anyscale 
https://github.com/ray-project/kuberay/pull/303#discussion_r921814738
I now save cluster name and job Id before continue reconciling rayJob. In this case, rayCluster will not be created duplicately because only the first updating of status will succeed

**Comment by harryge00** [comment]
@Jeffwan @brucez-anyscale Could you take another review? I have address the remaining issues

**Review by Jeffwan — APPROVED** [review]

File: ray-operator/controllers/ray/rayjob_controller.go
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/rayjob_controller.go:76]: If you update status in `setRayJobIdAndRayClusterNameIfNeed`, you should better just return this reconcile loop.
Or you should get `rayJobInstance` again, since the `resource version` of rayJobInstance has changed.

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (brucez-anyscale) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:126]: log.V(1).Info("fetchDashboardAgentURL ", "dashboardAgentURL", dashboardAgentURL)

Comment (Jeffwan) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:126]: @brucez-anyscale Seem this is the key change dashboardURL -> dashboardAgentURL? I can help update it in follow up PRs and let's merge this one. (We need this change in downstream to unblock other stories..)

**Pull request merged**
